java-frontend/src/main/java/org/sonar/java/resolve/TypeSubstitutionSolver.java

```diff
@@ -374,4 +390,52 @@ private boolean isValidSubstitution(TypeSubstitution candidate, TypeVariableJava
     return true;
   }
 
+  JavaType objectSubstitution(ParametrizedTypeJavaType type) {
+    TypeSubstitution substitution = new TypeSubstitution();
+    for (Map.Entry<TypeVariableJavaType, JavaType> entry : type.typeSubstitution.substitutionEntries()) {
+      TypeVariableJavaType typeVar = entry.getKey();
+      JavaType subs = entry.getValue();
+      if (typeVar == subs) {
+        substitution.add(typeVar, symbols.objectType);
+      } else {
+        substitution.add(typeVar, subs);
+      }
+    }
+    return parametrizedTypeCache.getParametrizedTypeType(type.symbol, substitution);
+  }
+
+  static TypeSubstitution substitutionFromSuperType(ParametrizedTypeJavaType target, ParametrizedTypeJavaType source) {
+    TypeSubstitution result = target.typeSubstitution.copy();
+    if (target.rawType != source.rawType) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Change this comparison to use the equals method. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1698)